manpages/uitest.man
.TH uitest 1 2/4/2009 BL3DEMC
.na
.nh

.SH NAME
uitest - a script to run GUI tests against Etomo using JFCUnit

.SH SYNOPSIS
.P
.B scripts/uitest \fITest_Section_name
.P
.B make -f tests/uitest.make \fR[\fItarget\fR ...]
.P
.B etomo.uitest.JfcUnitTests

.SH DESCRIPTION
\fBUitest\fR is a script that runs \fBJfcUnitTests\fR.  \fBJfcUnitTests\fR runs
the \fBetomo\fR application and interacts with it by using JfcUnit to grab
named GUI elements.  \fBEtomo\fR automatically names most of its GUI elements.
\fBJfcUnitTests\fR gets its instructions from a set of autodoc files which are
read sequentially.

.SH ENVIRONMENT
The IMOD environment that allows Etomo to be run must be in place.

.TP 4
.B IMOD_UITEST_DATA
Should be set to the directory where files that will be copied into a test
dataset are stored.

.TP 4
.B IMOD_UITEST_IMAGE_DATA
Should be set to the directory where images files that will be copied into a test
dataset are stored.

.TP 4
.B IMOD_UITEST_SOURCE
should be set the Etomo/tests directory.

.SS Used by the Uitest Script

.TP 4
.B IMOD_JUNIT_HOME
Should be set to the directory containing the JUnit jar file.

.TP 4
.B IMOD_JUNIT_JAR
Should be set to the name of the JUnit jar file.

.TP 4
.B IMOD_JFCUNIT_HOME
Should be set to the directory containing the JfcUnit jar file.

.TP 4
.B IMOD_JFCUNIT_JAR
Should be set to the name of the JfcUnit jar file.

.TP 4
.B IMOD_JAKARTA_REGEXP_HOME
Should be set to the directory containing the Jakarta regexp jar file.

.TP 4
.B IMOD_JAKARTA_REGEXP_JAR
Should be set to the Jakarta regexp jar file.

.SS Used in Uitest.make

.TP 4
.B IMOD_UITEST_SCRIPT
Should be set the directory containing uitest.

.SS Used by JfcUnitTests

.TP 4
.B IMOD_TEST_SECTION
Only needs to be set if \fBJfcUnitTests\fR is being run directly.  It should be
set to the name of \fBTest\fR section that will be run (see Section
Relationships below).

.SH FILES
.SS Autodocs

The main \fBuitest\fR autodoc file is \fBuitest.adoc\fR.  Test autodocs are
referred to by \fBuitest.adoc\fR.  Function autodocs are referred to by Test
autodocs.

.SH USAGE
.SS Section Relationships

The \fBuitest.adoc\fR file contains \fBTest\fR, \fBdataset\fR, and
\fBinterface\fR sections.  A \fBTest\fR section contains descriptions of tests
that will be run together.  A \fBTest\fR section refers to a single
\fBdataset\fR section for information on the dataset being tested.  Each
\fBdataset\fR section refers to one \fBinterface\fR section.  A \fBTest\fR
section also refers to at least one Test autodoc and specifies a section type to
be accessed in it.  Sections specified by a \fBTest\fR section are called Dialog
sections, and are accessed one after another.  Sections may also be called like
functions.  Subsections with a section type of \fBif\fR or \fBifnot\fR act as
simple if clauses.


.SS Testing
The TestRunner class picks a \fBTest\fR Section based on the
\fBIMOD_TEST_SECTION\fR environment variable (set by \fBuitest\fR).  Each
\fBTest\fR Section is associated with one dataset.  A list of autodocs are added
to the section using the \fBadoc\fR command.  Each autodoc is associated with
one Axis.  The \fBetomo.uitest.TestRunner\fR class runs a test on each autodoc
in turn, based on the order they appear in the \fBTest\fR Section.  Each test
will run until it starts to wait for a process, then it will return control to
\fBTestRunner\fR.  The \fBTestRunner\fR will keep running tests on each autodoc
in turn until all tests in the autodoc have been completed.  This allows
\fBuitest\fR to mimic the way a person might work with \fBEtomo\fR - moving from
one axis to another.


.SS Grammer

.TP 4
.I command
.I action \fB=\fR [\fIvalue\fR]
.RS 4

.P
.I action\fR[\fB.\fImodifier\fR]\fB.\fIfield \fB=\fR [\fIvalue\fR]

.P
.I action\fR[\fB.\fImodifier\fR]\fB.\fIfield.subcommand \fB=\fR [\fIvalue]

.P
.I action\fR[\fB.\fImodifier\fR]\fB.\fIsubject \fB=\fR [\fIvalue]

.P
.I action\fR[\fB.\fImodifier\fR]\fB.\fIsubject\fB.\fIfield \fB=\fR [\fIvalue]

.P
.I action\fR[\fB.\fImodifier\fR]\fB.\fIsubject\fB.\fIsubcommand \fB=\fR [\fIvalue]

.TP
.I field \fB=\fR [\fIvalue\fR]
A field command has an implied action; for instance the implied action for a
button (\fBbn\fR) is to push it.
.RE

.TP 4
.I subcommand
.RS

.P
.I subcommand_action\fR[\fB.\fImodifier\fR]\fB.\fIfield \fB=\fR [\fIvalue]

.P
.I subcommand_action\fR[\fB.\fImodifier\fR]\fB.\fIsubject \fB=\fR [\fIvalue]

.P
.I subcommand_action\fR[\fB.\fImodifier\fR]\fB.\fIsubject\fB.\fIfield \fB=\fR [\fIvalue]
.RE

.TP 4
.I action
.B assert\fR, \fBcopy\fR, \fBend\fR, \fBformat\fR, \fBif\fR, \fBgoto\fR, \fBopen\fR,
\fBreturn\fR, \fBrun\fR, \fBset\fR, \fBskipto\fR, \fBsleep\fR, \fBstart\fR, \fBtouch\fR,
\fBunset\fR, \fBuse\fR, \fBwait\fR, \fBwrite

.TP 4
.I subcommand_action
.B assert\fR, \fBreturn\fR, \fBrun\fR, \fBset\fR, \fBwait

.TP 4
.I modifier
.B always\fR, \fBcontains\fR, \fBdisabled\fR, \fBenabled\fR,
\fBequals\fR, \fBexists\fR, \fBge\fR, \fBkeep\fR, \fBle\fR, \fBnot\fR,
\fBnot-equals\fR, \fBnot-exists\fR, \fBsame\fR, \fBsingle\fR, \fBwait

.TP 4
.I subject
.I subject_type\fR[\fB.\fIsubject_name\fR][\fB.\fIaxis\fR]
.RS 4

.TP 4
.I subject_type
.B adoc\fR, \fBdataset\fR, \fBdebug\fR, \fBdialog\fR, \fBdir\fR, \fBfile\fR,
\fBfile-chooser\fR, \fBframe\fR, \fBfunction\fR, \fBindex\fR, \fBinterface\fR, \fBparam\fR, \fBpopup\fR, \fBprocess\fR, \fBsection\fR, \fBtest\fR, \fBtestdir\fR,
\fBvar

.TP 4
.I subject_name
Name of the subject.  May contain variables.  May not be \fBa\fR, \fBb\fR, an
\fIsubcommand_action\fR string, or a \fIfield\fR string.

.TP 4
.I axis
.B a \fRor \fBb\fR.  Default is single axis.
.RE

.TP 4
.I field
.I field_type\fB.\fIfield_name\fR[\fB.\fIindex\fR]
.RS 4

.P
.I field_type
.RS 4

.TP 4
.B bn
A button; the implied action is to press the button.

.TP 4
.B cb
A check box; the implied action is to check the check box.

.TP 4
.B cbb
A combo box; the implied action is to add an item to the combo box and select it.

.TP 4
.B mb
A mini-button; the implied action is to press the mini-button.

.TP 4
.B pnl
A panel; the implied action is to go to the panel.

.TP 4
.B sp
A spinner; the implied action is to either enter a number into the spinner's text

History
IMOD History
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
7/3/13: In eTomo improving error handling for an empty fiducial diameter
field.(Bug# 1671)

1/20/13: IMOD_4-6-38
1/20/13: Fixed keypad hot keys in Model View on Mac.

1/18/14: IMOD_4-6-37
1/15/14: Fixed problems with autofidseed in eTomo - Fiducial Model Dialog.
(Bug# 1743)

1/15/14: Exposed flgNoReferenceRefinement in eTomo - PEET.(Bug# 1738)

1/18/14: Further polishing of Grab with Note, which can now be used to draw
multiple arrows in Zap or Slicer windows.

1/16/14: IMOD_4-6-36
1/16/14: Jane Ding contributed Grab with Note plugin, which takes Slicer or Zap
snapshots with notes added below the image.

1/16/14: Fixed creation of patch tracking com file if mag change searched in
coarse align.

1/16/14: Change eTomo to make it independent of Cygwin.  Fixed a problem with
killing processes on Windows.(Bug# 1716)

1/14/14: Fixed bug in clip multiplying while unpacking 4-bit data, added
rounding in clip multiply or divide when outout is to an integer.

1/12/14: Made Imodchopconts able to chop up or sort out contours based on
fine-grained color.

1/11/14: Fixed 3dmod crash when the I button is pressed in Zap and no image
loaded, present since 4.6.29.

1/10/14: Made 3dmod able to read single pixels from non-MRC-like files so
Pixel View and Graph windows can show file values.
 
1/8/14: Made all C/C++ programs except Raptor able to read TIFF and MRC-like
file formats, made it possible to direct output to a format specified by
environment variable. TIFF stacks are now written with min and max of the file
in the last image directory and these values are used when reading a stack.

1/8/14: Added control of imodchopconts to eTomo - Fiducial Model Dialog.
(Bug# 1767)

1/2/14: Replaced Fortran image I/O system with system used by C programs.
Fortran programs can now read TIFF, DigitalMicrograph, and other MRC-like
files.

1/2/14: IMOD_4-6-35
1/2/14: Fixes in 3dmod for problems with toolbar and popup menu and new model
view tool windows.

12/31/13: IMOD_4-6-34
12/25/13: Changed description of hot keys in help files and tooltips to show
plain letter as upper instead of lower case (S not s) and shifted letter with
Shift+ instead or just upper case (Shift+S not S).

12/23/13: Added ability to change object transparency during Model View
movies. 

12/23/13: Made hot key popup menus for Zap, Slicer, and XYZ windows that show
up when right-clicking over a toolbar, and added a HotKey menu to Model View.
These menus allow most hot keys to be seen and selected.

12/23/13: Added hot key for resizing image inside Zap or Slicer rubber band to
fill window.

12/22/13: Added a dialog to set the size of the Model View window without
changing model zoom.
 
12/22/13: Added menu entry Image - Linked Slicers to open an array of slicers,
one per  image file, controlled by a single slice toolbar,

12/21/13: Added a small rotation tool window to Model View with arrows for 
rotating model.

12/19/13: Rubberband, arrows for view-axis based rotation, and slider for

Etomo/src/etomo/uitest/UITest.java
package etomo.uitest;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.util.ArrayList;

import etomo.Arguments;
import etomo.EtomoDirector;
import etomo.process.SystemProgram;
import etomo.storage.LogFile;
import etomo.storage.autodoc.AutodocFactory;
import etomo.storage.autodoc.ReadOnlyAutodoc;
import etomo.storage.autodoc.ReadOnlyStatement;
import etomo.storage.autodoc.ReadOnlySection;
import etomo.type.AxisID;
import etomo.type.UITestAction;
import etomo.ui.UIHarness;
import etomo.util.EnvironmentVariable;
import etomo.util.Utilities;
import junit.extensions.jfcunit.JFCTestCase;
import junit.extensions.jfcunit.JFCTestHelper;

/**
 * <p>Description: </p>
 * 
 * <p>Copyright: Copyright (c) 2005</p>
 *
 * <p>Organization:
 * Boulder Laboratory for 3-Dimensional Electron Microscopy of Cells (BL3DEM),
 * University of Colorado</p>
 * 
 * @author $Author$
 * 
 * @version $Revision$
 */
public final class UITest extends JFCTestCase implements UITestCommandFactory {
  public static final String rcsid = "$Id$";

  static final File SOURCE_DIR = Utilities.getExistingDir("IMOD_UITEST_SOURCE",
      AxisID.ONLY);
  private static final File DATA_DIR = Utilities.getExistingDir(
      "IMOD_UITEST_DATA", AxisID.ONLY);
  private static final String[] ARGS = new String[] { Arguments.SELFTEST_TAG,
      Arguments.TEST_TAG/*, Arguments.NAMES_TAG*/ };
  static final long DEFAULT_SLEEP = 1000;

  private static Throwable uncaughtException = null;

  private CommandReader reader = null;
  private File datasetDir = null;
  private ReadOnlyAutodoc autodocA = null;
  private ReadOnlyAutodoc autodocB = null;
  private JFCTestHelper helper = null;
  private EtomoDirector etomo = null;
  private long sleep = DEFAULT_SLEEP;
  private String dataset = null;
  private boolean keep = false;
  private UITestAxis axisAUITest = null;
  private UITestAxis axisBUITest = null;
  private long duration = 15 * 60;//test duration in seconds
  private File dataDir = null;
  private AxisID axisIDA = null;
  private AxisID axisIDB = null;
  private String dataFileName = null;
  private ArrayList variablesA = null;
  private ArrayList variablesB = null;
  private InterfaceSection interfaceSection = null;

  protected void setUp() throws Exception {
    super.setUp();
    helper = new JFCTestHelper();
    setHelper(helper);
  }

  protected void tearDown() throws Exception {
    if (etomo != null) {
      etomo.setTestDone(true);
      etomo.exitProgram(AxisID.ONLY);
      etomo = null;
    }
    JFCTestHelper.cleanUp(this);
    super.tearDown();
  }

  /**
   * run all the tests in uitest.adoc
   * @throws IOException
   */
  public void test() throws IOException, LogFile.ReadException,
      LogFile.FileException {
    String testName = EnvironmentVariable.INSTANCE.getValue(null,
        "IMOD_TEST_SECTION", AxisID.ONLY);
    System.err.println("test " + testName + ":");
    if (testName == null || testName.matches("\\*+")) {
      fail("$IMOD_TEST_SECTION has not been set");
    }
    //get the uitest.adoc
    ReadOnlyAutodoc autodoc = null;
    try {
      autodoc = AutodocFactory.getInstance(AutodocFactory.UITEST, AxisID.ONLY);
    }
    catch (FileNotFoundException e) {
      fail(e.getMessage());
      return;
    }
    catch (IllegalStateException e) {
      e.printStackTrace();
      fail(e.getMessage());
    }
    if (autodoc == null) {
      fail("Missing autodoc");
      return;
    }
    reader = new CommandReader(autodoc, TestSectionCommand.SECTION_TYPE);
    CommandFactory factory = new CommandFactory();
    UITestAutodocCommand command = (UITestAutodocCommand) reader.nextCommand(
        null, factory);
    //process commands in the global section
    while (!command.isEmpty()) {
      if (command.isKnown()) {
        UITestAction action = command.getAction();
        if (action == UITestAction.SLEEP) {
          setSleep(command);
        }
      }
      command = (UITestAutodocCommand) reader.nextCommand(command, factory);
    }
    //process section specified by testName
    reader.setSection(testName);
    if (reader.isDone()) {
      return;
    }
    processSection(autodoc);
    UIHarness.INSTANCE.setLog(true);
    runTests();
  }

  private void setSleep(UITestAutodocCommand command) {
    String value = command.getValue();
    assertNotNull(reader.getInfo(), "Unknown name/value pair format: "
        + command, value);
    try {
      sleep = Long.parseLong(value);
    }
    catch (NumberFormatException e) {
      fail(reader.getInfo(), "value must be numeric: " + command);
    }
  }

  void moveSubFrame() {
    UIHarness.INSTANCE.moveSubFrame();
  }

  long getSleep() {
    return sleep;
  }

  String getDataset() {
    return dataset;
  }

  public UITestCommand newCommand() {
    return new TestSectionCommand();
  }

  private void setDuration(TestSectionCommand command) {
    String[] durationArray = command.getValue().split(":");
    if (durationArray.length == 4 && !durationArray[3].equals("")) {
      duration = Long.parseLong(durationArray[3]);
    }
    else {
      duration = 0;
    }
    if (durationArray.length >= 3 && !durationArray[2].equals("")) {
      duration += Long.parseLong(durationArray[2]) * 60;
    }
    if (durationArray.length >= 2 && !durationArray[1].equals("")) {
      duration += Long.parseLong(durationArray[1]) * 60 * 60;
    }
    if (durationArray.length >= 1 && !durationArray[0].equals("")) {
      duration += Long.parseLong(durationArray[0]) * 60 * 60 * 24;
    }
  }

  private String getRequiredValue(TestSectionCommand command) {
    String value = command.getValue();
    assertNotNull(reader.getInfo(), "Unknown name/value pair format: "
        + command, value);
    return value;
  }

  private void setVariable(TestSectionCommand command) {
    AxisID axisID = command.getAxisID();
    if (axisID == null || axisID == AxisID.ONLY || axisID == AxisID.FIRST) {
      variablesA = setVariable(command, variablesA);
    }
    if (axisID == null || axisID == AxisID.SECOND) {
      variablesB = setVariable(command, variablesB);
    }
  }

  private ArrayList setVariable(TestSectionCommand command, ArrayList variables) {
    if (variables == null) {
      variables = new ArrayList();
    }
    variables.add(command.getVariable());
    return variables;
  }

  private void setDataset(TestSectionCommand command) {
    String value = command.getValue();
    assertNotNull(reader.getInfo(), "Unknown name/value pair format: "
        + command, value);
    dataset = value;
    setVariable(command);
  }

  /**
   * Performs all commands and open all autodocs found in the section
   * @throws FileNotFoundException
   * @throws IOException
   */
  private void processSection(ReadOnlyAutodoc autodoc)
      throws FileNotFoundException, IOException, LogFile.ReadException,
      LogFile.FileException {
    TestSectionCommand command = (TestSectionCommand) reader.nextCommand(null,
        this);
    while (!command.isEmpty()) {
      if (command.isKnown()) {
        UITestAction action = command.getAction();
        if (action == UITestAction.ADOC) {
          setAutodoc(command);
        }
        else if (action == UITestAction.COPY) {
          if (datasetDir == null) {
            setDatasetDir(null, false);
          }
          copyFile(command);
        }
        else if (action == UITestAction.DATA_FILE) {
          dataFileName = getRequiredValue(command);
        }
        else if (action == UITestAction.DATASET) {
          setDataset(command);
        }
        else if (action == UITestAction.DATASET_DIR) {
          if (datasetDir == null) {
            setDatasetDir(command);
          }
        }
        else if (action == UITestAction.DURATION) {
          setDuration(command);
        }
        else if (action == UITestAction.DATA_DIR) {
          //Get a file directory relative to the base file dir
          dataDir = getRelativeDir(command, DATA_DIR, false);
        }
        else if (action == UITestAction.SET) {
          setVariable(command);
        }
        else if (action == UITestAction.INTERFACE) {
          interfaceSection = InterfaceSection.getInstance(autodoc, command);
        }
      }
      command = (TestSectionCommand) reader.nextCommand(command, this);
    }
    assertTrue(reader.getInfo(), UITestAction.ADOC + " is required.",
        autodocA != null || autodocB != null);
    if (datasetDir == null) {
      setDatasetDir(null, false);
    }
  }

  long getDuration() {
    return duration;
  }

  /**
   * runs tests with uitest-axis autodocs for A and/or B
   * @param test
   * @param sourceDir
   */
  private void runTests() {
    if (autodocA == null && autodocB == null) {
      return;
    }
    //run Etomo
    if (dataFileName == null) {
      EtomoDirector.main(ARGS);
    }
    else {
      File dataFile = new File(System.getProperty("user.dir"), dataFileName);
      String[] args = new String[ARGS.length + 1];
      for (int i = 0; i < ARGS.length; i++) {
        args[i] = ARGS[i];
      }
      args[ARGS.length] = dataFile.getAbsolutePath();
      EtomoDirector.main(args);
    }
    etomo = EtomoDirector.INSTANCE;
    //Create the UITestAxis instances
    if (autodocA != null) {
      axisAUITest = new UITestAxis(this, autodocA, helper, axisIDA,
          dataFileName != null, variablesA, interfaceSection);
    }
    if (autodocB != null) {
      axisBUITest = new UITestAxis(this, autodocB, helper, axisIDB,
          dataFileName != null, variablesB, interfaceSection);
    }
    //Test the axis or axes, taking turns if there are two
    boolean testingA = axisAUITest != null;
    boolean testingB = axisBUITest != null;
    while ((testingA && !axisAUITest.isDone())
        || (testingB && !axisBUITest.isDone())) {
      if (testingA && !axisAUITest.isDone()) {
        axisAUITest.testAxis();
      }
      if (testingB && !axisBUITest.isDone()) {
        axisBUITest.testAxis();
      }
    }
  }

  /**
   * Returns true if the axis test is null.
   * Returns true if the axis test is done and not stopped.
   * Otherwise returns true if able to remove the dialog from the finished
   * dialog list.
   * @param axisID
   * @param dialogType
   * @return
   */
  boolean removeDialog(AxisID axisID, String dialogSectionName) {
    UITestAxis axisUITest;
    if (axisID == AxisID.SECOND) {
      axisUITest = axisBUITest;
    }
    else {
      axisUITest = axisAUITest;
    }
    if (axisUITest == null) {
      return true;
    }
    return axisUITest.removeFinishedDialog(dialogSectionName);
  }

  /**
   * returns true if the axis test exists and is stopped
   * @param axisID
   * @return
   */
  boolean isStopped(AxisID axisID) {
    UITestAxis axisUITest;
    if (axisID == AxisID.SECOND) {
      axisUITest = axisBUITest;
    }
    else {
      axisUITest = axisAUITest;
    }
    if (axisUITest == null) {
      return false;
    }
    return axisUITest.isStopped();
  }

  /**
   * copy a file from dataDir to the working directory
   * @param attrib
   * @param sourceDir
   */
  void copyFile(UITestCommand command) {
    assertNotNull(command.toString()
        + ":  The dataDir must be set before a copy can be done.");
    String fileName = command.getValue();
    assertNotNull(reader.getInfo(), "Unknown name/value pair format: "
        + command, fileName);
    File file = new File(dataDir, fileName);
    if (!file.exists() || file.isDirectory()) {
      throw new IllegalStateException("bad dataset file: "
          + file.getAbsolutePath());
    }
    if (keep) {
      File targetFile = new File(System.getProperty("user.dir"), file.getName());
      if (targetFile.exists()) {
        return;
      }
    }
    SystemProgram copy = new SystemProgram(System.getProperty("user.dir"),
        new String[] { "cp", file.getAbsolutePath(), "." }, AxisID.ONLY);
    copy.run();
  }

  /**
   * Create a new autodoc and set it to autodocA or B, if the autodocA(B) is not
   * already set. 
   * @param adocAttrib
   * @param currentSourceDir
   * @throws FileNotFoundException
   * @throws IOException
   */
  private void setAutodoc(TestSectionCommand command)
      throws FileNotFoundException, IOException, LogFile.ReadException,
      LogFile.FileException {
    String value = command.getValue();
    assertNotNull(null, "Unknown name/value pair format: " + command, value);
    setVariable(command);
    //look for adoc = autodoc
    AxisID axisID = command.getAxisID();
    if (axisID == AxisID.SECOND) {
      if (autodocB != null) {
        return;
      }
      axisIDB = axisID;
      autodocB = AutodocFactory.getInstance(SOURCE_DIR, value, AxisID.ONLY);
      return;
    }
    if (autodocA != null) {
      return;
    }
    axisIDA = axisID;
    autodocA = AutodocFactory.getInstance(SOURCE_DIR, value, AxisID.ONLY);
  }

  /**
   * Get the datasetDir for the current section.  If the keep is false 
   * then clean the datasetDir by deleting
   * it and then recreating it.  Set it be the working directory.
   * @param datasetDir
   * @return
   */
  private void setDatasetDir(String dirName, boolean keep) {
    assertNull("The dataset directory can only be set once:dirName=" + dirName
        + ",datasetDir=", datasetDir);
    JfcUnitTests.TEST_ROOT_DIR.mkdirs();
    //Get the directory name
    if (dirName == null) {
      dirName = reader.getName();
      assertNotNull(reader.getInfo(), "Invalid section", dirName);
    }
    //make the directory path
    datasetDir = new File(JfcUnitTests.TEST_ROOT_DIR, dirName);
    if (!datasetDir.exists()) {
      datasetDir.mkdirs();
    }
    //if keep, do not clean the directory
    if (!keep) {
      //clean the directory by deleting it
      SystemProgram remove = new SystemProgram(System.getProperty("user.dir"),
          new String[] { "rm", "-fr", datasetDir.getAbsolutePath() },
          AxisID.ONLY);
      remove.run();
      //make the directory
      datasetDir.mkdir();
    }
    //make the directory the working directory
    System.setProperty("user.dir", datasetDir.getAbsolutePath());
  }

  private void setDatasetDir(TestSectionCommand command) {
    File datasetDir = null;
    //Get the directory name
    String dirName = command.getValue();
    boolean keep = command.isKeep();
    setDatasetDir(dirName, keep);
  }

  private File getAbsoluteDir(UITestCommand command, boolean makeDirs) {
    //Get the directory name
    String dirName = command.getValue();
    assertNotNull(reader.getInfo(), "Unknown name/value pair format: "
        + command, dirName);
    //make the directories
    File dir = new File(dirName);
    assertTrue(reader.getInfo(), dir.isAbsolute());
    if (makeDirs && !dir.exists()) {
      dir.mkdirs();
    }
    return dir;
  }

  /**
   * Create a directory or directory path in rootDir using a command value.
   * If the value does not exist, return rootDir.  Make
   * directories if makeDirs is true and the directory does not exist.
   * This is an optional attribute:  return rootDir if the attribute is not found.
   * @param autodoc
   * @param attribName
   * @param rootDir
   * @param makeDirs - if true, make the directories
   * @return
   */
  private File getRelativeDir(UITestCommand command, File rootDir,
      boolean makeDirs) {
    //Make the root directories on the root directory since it could be returned
    if (makeDirs && !rootDir.exists()) {
      rootDir.mkdirs();
    }
    //Get the directory name
    String dirName = command.getValue();
    assertNotNull(reader.getInfo(), "Unknown name/value pair format: "
        + command, dirName);
    //make the directories
    File dir = new File(rootDir, dirName);
    if (makeDirs && !dir.exists()) {
      dir.mkdirs();
    }
    return dir;
  }

  /**
   * Create a directory or directory path in rootDir using an attribute value.
   * If the attribute value does not exist, return originalDir.  Make
   * directories if makeDirs is true and the directory does not exist.
   * @param section
   * @param attribName
   * @param rootDir
   * @return the directory
   */
  private File getRelativeDir(ReadOnlyStatement pair, File rootDir,
      boolean makeDirs) {
    //Get the directory name
    String dirName = pair.getRightSide();
    assertNotNull(null, "Unknown name/value pair format: " + pair.getString(),
        dirName);
    //make the directories
    File dir = new File(rootDir, dirName);
    if (makeDirs && !dir.exists()) {
      dir.mkdirs();
    }
    return dir;
  }

  /**
   * Create a File in rootDir using the section name.  The section name is a
   * required field so this function should never return rootDir.
   * @param section
   * @param rootDir
   * @param makeDirs
   * @return
   */
  private File getRequiredRelativeDir(ReadOnlySection section, File rootDir,
      boolean makeDirs) {
    //Get the directory name
    String dirName = section.getName();
    //make the directories
    File dir = new File(rootDir, dirName);
    if (makeDirs && !dir.exists()) {
      dir.mkdirs();
    }
    return dir;
  }

  public void assertTrue(String sectionInfo, String message, boolean condition) {
    if (sectionInfo != null) {
      message = sectionInfo + ": " + message;
    }
    assertTrue(message, condition);
  }

  public void fail(String sectionInfo, String command, String message) {
    StringBuffer buffer = new StringBuffer();
    if (sectionInfo != null) {
      buffer.append(sectionInfo + ": ");
    }
    if (command != null) {
      buffer.append(command + ":\n");
    }
    buffer.append(message);
    fail(message);
  }

  public void fail(String sectionInfo, String message) {
    if (sectionInfo != null) {
      message = sectionInfo + ": " + message;
    }
    fail(message);
  }

  public void assertNotNull(String sectionInfo, String message, Object object) {
    if (sectionInfo != null) {
      message = sectionInfo + ": " + message;
    }
    assertNotNull(message, object);
  }

  public void assertFalse(String sectionInfo, String message, boolean condition) {
    if (sectionInfo != null) {
      message = sectionInfo + ": " + message;
    }
    assertFalse(message, condition);
  }

  public void assertNull(String sectionInfo, String message, Object object) {
    if (sectionInfo != null) {
      message = sectionInfo + ": " + message;
    }
    assertNull(message, object);
  }

  public void assertNull(String message, File file) {
    if (file == null) {
      return;
    }
    fail(message + file.getAbsolutePath());
  }

  public void assertEquals(String sectionInfo, String message, Object expected,
      Object actual) {
    if (sectionInfo != null) {
      message = sectionInfo + ": " + message;
    }
    assertEquals(message, expected, actual);
  }

  public void assertEquals(String sectionInfo, String message,
      boolean expected, boolean actual) {
    if (sectionInfo != null) {
      message = sectionInfo + ": " + message;
    }
    assertEquals(message, expected, actual);
  }

  private static final class CommandFactory implements UITestCommandFactory {
    public UITestCommand newCommand() {
      return new UITestAutodocCommand();
    }
  }
}
/**
 * <p> $Log$
 * <p> Revision 1.2  2008/05/30 22:39:56  sueh
 * <p> bug# 1102 Isolating the etomo.uitest package so it is not needed for
 * <p> running EtomoDirector.
 * <p>
 * <p> Revision 1.1  2008/05/30 21:43:40  sueh
 * <p> bug# 1102 Moved uitest classes to etomo.uitest.  Added InterfaceSection.
 * <p>
 * <p> Revision 1.29  2008/01/31 20:31:32  sueh
 * <p> bug# 1055 throwing a FileException when LogFile.getInstance fails.
 * <p>
 * <p> Revision 1.28  2008/01/25 18:50:30  sueh
 * <p> bug# 1069 Changed setDuration to be more flexible about the format it can read.
 * <p>
 * <p> Revision 1.27  2007/12/26 22:40:03  sueh
 * <p> bug# 1052 Moved argument handling from EtomoDirector to a separate class.
 * <p>
 * <p> Revision 1.26  2007/09/07 00:30:03  sueh
 * <p> bug# 989 Using a public INSTANCE to refer to the EtomoDirector singleton
 * <p> instead of getInstance and createInstance.
 * <p>
 * <p> Revision 1.25  2007/04/09 21:23:53  sueh
 * <p> Bug# 964 Changed NameValuePair to Statement and child classes.
 * <p>
 * <p> Revision 1.24  2007/03/21 19:48:12  sueh
 * <p> bug# 964 Limiting access to autodoc classes by using ReadOnly interfaces.
 * <p> Added AutodocFactory to create Autodoc instances.  Moved AdocCommand
 * <p> classes out of the autodoc package.
 * <p>
 * <p> Revision 1.23  2007/03/05 21:29:27  sueh
 * <p> bug# 964 Stop controlling autodoc instances, except for the standard ones.
 * <p>
 * <p> Revision 1.22  2007/03/01 01:45:51  sueh
 * <p> bug# 964 Added LogFile to Autodoc.
 * <p>
 * <p> Revision 1.21  2006/10/24 23:34:49  sueh
 * <p> bug# 948 Changed filedir to datadir.  Removed testdir.
 * <p>
 * <p> Revision 1.20  2006/10/11 10:12:33  sueh
 * <p> bug# 938 Making ThreadGroup independent of UITest, so that it does not require
 * <p> JfcUnit to compile.
 * <p>
 * <p> Revision 1.19  2006/10/10 05:25:25  sueh
 * <p> bug# 931 Added uncaughtException to handle uncaught exceptions in Etomo
 * <p> during UITest.
 * <p>
 * <p> Revision 1.18  2006/08/28 18:27:04  sueh
 * <p> bug# 923 Changed the source attribute to filedir.  Global filedir is an absolute file
 * <p> path.
 * <p>
 * <p> Revision 1.17  2006/06/30 16:29:54  sueh
 * <p> bug# 883 Added EnvironmentVariable, a class to get and store environment
 * <p> variables.
 * <p>
 * <p> Revision 1.16  2006/06/14 00:41:33  sueh
 * <p> bug# 852 Moved classes to the autodoc package that parse an autodoc or find
 * <p> attributes specific to a type of autdoc.
 * <p>
 * <p> Revision 1.15  2006/06/06 18:16:19  sueh
 * <p> bug# 766 Fail when test name is not set.  Setting log in UIHarness.
 * <p>
 * <p> Revision 1.14  2006/06/06 17:21:45  sueh
 * <p> bug# 766 Add fails when autodoc is not found
 * <p>
 * <p> Revision 1.13  2006/05/11 19:33:00  sueh
 * <p> Removed unnessary strings from ARGS.
 * <p>
 * <p> Revision 1.12  2006/05/01 21:19:43  sueh
 * <p> bug# 787 Removed fiducial diameter, added set.
 * <p>
 * <p> Revision 1.11  2006/04/28 21:06:48  sueh
 * <p> bug# 787 Simplify code by using AdocCommandReader instead of
 * <p> keyed searches and NameValuePair.
 * <p>
 * <p> Revision 1.10  2006/04/25 19:38:45  sueh
 * <p> bug# 787 Added duration and datafile.  Added removeDialog() to remove
 * <p> a done dialog, when checking for a completed dialog.
 * <p>
 * <p> Revision 1.9  2006/01/13 20:17:05  sueh
 * <p> bug# 675 Added datasetdir.keep to avoid copying the dataset files.
 * <p>
 * <p> Revision 1.8  2006/01/12 22:18:18  sueh
 * <p> bug# 675 removed unnecessary function enterClickAndLeave().
 * <p>
 * <p> Revision 1.7  2006/01/12 17:38:44  sueh
 * <p> bug# 798 Moved the autodoc classes to etomo.storage.autodoc.
 * <p>
 * <p> Revision 1.6  2006/01/11 23:18:40  sueh
 * <p> bug# 675 Running one test instead of all tests.  Enforcing uitest.adoc
 * <p> syntax rules.  Added fiducial diameter.
 * <p>
 * <p> Revision 1.5  2006/01/04 20:28:12  sueh
 * <p> bug# 675 Moved constants that must be shared by non-test objects to an
 * <p> object which doesn't know about junit.  Overwise junit would have to be in
 * <p> the path for compiling and running EtomoDirector.
 * <p>
 * <p> Revision 1.4  2006/01/04 00:08:44  sueh
 * <p> bug# 675 Make test() run one test described in uitest.adoc.
 * <p>
 * <p> Revision 1.3  2005/12/30 21:19:57  sueh
 * <p> bug# 675 class to run a jfcunit test
 * <p>
 * <p> Revision 1.1  2005/12/23 02:25:03  sueh
 * <p> bug# 675 A class to generically run Etomo through the UI.
 * <p> </p>
 */
